includes/razorpay-subscriptions.php

```diff
@@ -0,0 +1,372 @@
+
+        $subscriptionData = $this->getSubscriptionCreateData($orderId);
+
+        try
+        {
+            $subscription = $this->api->subscription->create($subscriptionData);
+        }
+        catch (Exception $e)
+        {
+            $message = $e->getMessage();
+
+            throw new Errors\Error(
+                $message,
+                WooErrors\ErrorCode::API_SUBSCRIPTION_CREATION_FAILED,
+                400
+            );
+        }
+
+        // Setting the subscription id as the session variable
+        $sessionKey = $this->getSubscriptionSessionKey($orderId);
+        $woocommerce->session->set($sessionKey, $subscription['id']);
+
+        return $subscription['id'];
+    }
+
+    public function cancelSubscription($subscriptionId)
+    {
+        try
+        {
+            $subscription = $this->api->subscription->cancel($subscriptionId);
+        }
+        catch (Exception $e)
+        {
+            $message = $e->getMessage();
+
+            throw new Errors\Error(
+                $message,
+                WooErrors\ErrorCode::API_SUBSCRIPTION_CANCELLATION_FAILED,
+                400
+            );
+        }
+    }
+
+    protected function getSubscriptionCreateData($orderId)
+    {
+        $order = new WC_Order($orderId);
+
+        $product = $this->getProduct($order);
+
+        $planId = $this->getProductPlanId($product);
+
+        $customerId = $this->getCustomerId($order);
+
+        $length = (int) WC_Subscriptions_Product::get_length($product['product_id']);
+
+        $subscriptionData = array(
+            'customer_id'     => $customerId,
+            'plan_id'         => $planId,
+            'quantity'        => (int) $product['qty'],
+            'total_count'     => $length,
+            'customer_notify' => 0,
+            'notes'           => array(
+                'woocommerce_order_id'   => $orderId,
+                'woocommerce_product_id' => $product['product_id']
+            ),
+        );
+
+        $signUpFee = WC_Subscriptions_Product::get_sign_up_fee($product['product_id']);
+
+        if ($signUpFee)
+        {
+            $item = array(
+                'amount'   => (int) round($signUpFee * 100),
+                'currency' => get_woocommerce_currency(),
+                'name'     => $product['name']
+            );
+
+            if ($item['currency'] !== self::INR)
+            {
+                $this->razorpay->handleCurrencyConversion($item);
+            }
+
+            $subscriptionData['addons'] = array(array('item' => $item));
+        }
+
+        return $subscriptionData;
+    }
+
+    protected function getProductPlanId($product)
+    {
+        $currency = get_woocommerce_currency();
+
+        $key = self::RAZORPAY_PLAN_ID . '_'. strtolower($currency);
+
+        $productId = $product['product_id'];
+
+        $metadata = get_post_meta($productId);
+
+        list($planId, $created) = $this->createOrGetPlanId($metadata, $product);
```

Comment:
suggestion: pass currency to this method, so that we don't have to fetch it again. Makes it more deterministic.